File: packages/android/src/main/java/com/kozmos/components/Counter/Counter.kt
package com.kozmos.components.counter

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.widthIn
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.kozmos.tokens.KozmosDimensions
import com.kozmos.components.KozmosEmotion
import com.kozmos.tokens.KozmosThemeTokens

enum class CounterTone {
    Neutral,
    Brand,
    Destructive,
    Inverse
}

enum class CounterSize {
    Default,
    Sm
}

@Composable
fun KozmosCounter(
    text: String,
    modifier: Modifier = Modifier,
    tone: CounterTone = CounterTone.Neutral,
    size: CounterSize = CounterSize.Default,
    /**
     * Unset and `tone` draws as it always has; set and the emotion decides the
     * colour. A counter is always a filled pill, so there is one treatment.
     */
    emotion: KozmosEmotion? = null,
    /** A fill of the host's own — a category's colour — over the tone's and
     *  the emotion's; the digits go white on it. */
    fill: Color? = null
) {
    val height = if (size == CounterSize.Sm) 18.dp else 20.dp
    val minWidth = if (size == CounterSize.Sm) 18.dp else 20.dp
    val horizontalPadding = if (size == CounterSize.Sm) 5.dp else KozmosDimensions.primitivesLayoutSpacing75
    val fontSize = if (size == CounterSize.Sm) 11.sp else 12.sp

    val containerColor = when (tone) {
        CounterTone.Neutral -> KozmosThemeTokens.componentsPrimaryButtonsNeutralButtonBackgroundIdle
        CounterTone.Brand -> KozmosThemeTokens.componentsPrimaryButtonsThemedButtonBackgroundIdle
        CounterTone.Destructive -> KozmosThemeTokens.componentsPrimaryButtonsDangerButtonBackgroundIdle
        CounterTone.Inverse -> KozmosThemeTokens.primitivesColorsBackground0
    }

    val contentColor = when (tone) {
        CounterTone.Neutral -> KozmosThemeTokens.componentsPrimaryButtonsNeutralButtonForegroundContentIdle
        CounterTone.Brand -> KozmosThemeTokens.componentsPrimaryButtonsThemedButtonForegroundContentIdle
        CounterTone.Destructive -> KozmosThemeTokens.componentsPrimaryButtonsDangerButtonForegroundContentIdle
        CounterTone.Inverse -> KozmosThemeTokens.primitivesColorsForeground100
    }

    val resolvedContainer = fill ?: emotion?.surface ?: containerColor
    val resolvedContent = if (fill != null) KozmosThemeTokens.primitivesColorsBackground0 else (emotion?.onSurface ?: contentColor)

    Box(
        modifier = modifier
            .height(height)
            .widthIn(min = minWidth)
            .background(resolvedContainer, CircleShape)
            .padding(horizontal = horizontalPadding),
        contentAlignment = Alignment.Center
    ) {
        Text(
            text = normalizeCounterText(text),
            color = resolvedContent,
            fontSize = fontSize,
            fontWeight = FontWeight.SemiBold,
            maxLines = 1
        )
    }
}

private fun normalizeCounterText(value: String): String {
    val trimmed = value.trim()
    if (trimmed.startsWith("(") && trimmed.endsWith(")")) {
        return trimmed.removePrefix("(").removeSuffix(")").trim()
    }

    return trimmed
}

File: packages/ios/Sources/Components/Counter/Counter.swift
import SwiftUI

public enum KozmosCounterTone {
    case neutral
    case brand
    case destructive
    case inverse
}

public enum KozmosCounterSize {
    case `default`
    case sm
}

public struct KozmosCounter: View {
    let text: String
    let tone: KozmosCounterTone
    let size: KozmosCounterSize
    /// Unset and `tone` draws as it always has; set and the emotion decides
    /// the colour. A counter is always a filled pill, so there is one
    /// treatment rather than two.
    let emotion: KozmosEmotion?
    /// A fill of the host's own — a category's colour — over the tone's and
    /// the emotion's; the digits go white on it.
    let fill: Color?

    public init(
        _ text: String,
        tone: KozmosCounterTone = .neutral,
        size: KozmosCounterSize = .default,
        emotion: KozmosEmotion? = nil,
        fill: Color? = nil
    ) {
        self.text = text
        self.tone = tone
        self.size = size
        self.emotion = emotion
        self.fill = fill
    }

    public var body: some View {
        Text(normalizedText)
            .font(.system(size: fontSize, weight: .semibold))
            .monospacedDigit()
            .lineLimit(1)
            .padding(.horizontal, horizontalPadding)
            .frame(minWidth: minWidth, minHeight: height)
            .background(fill ?? emotion?.surface ?? backgroundColor)
            .foregroundColor(fill != nil ? KozmosColors.primitivesColorsBackground0 : (emotion?.onSurface ?? foregroundColor))
            .clipShape(Capsule())
    }

    private var normalizedText: String {
        let trimmed = text.trimmingCharacters(in: .whitespacesAndNewlines)
        if trimmed.hasPrefix("("), trimmed.hasSuffix(")") {
            return String(trimmed.dropFirst().dropLast()).trimmingCharacters(in: .whitespacesAndNewlines)
        }

        return trimmed
    }

    private var height: CGFloat {
        size == .sm ? 18 : 20
    }

    private var minWidth: CGFloat {
        size == .sm ? 18 : 20
    }

    private var horizontalPadding: CGFloat {
        size == .sm ? 5 : KozmosDimensions.primitivesLayoutSpacing75
    }

    private var fontSize: CGFloat {
        size == .sm ? 11 : 12
    }

    private var backgroundColor: Color {
        switch tone {
        case .neutral:
            return KozmosColors.componentsPrimaryButtonsNeutralButtonBackgroundIdle
        case .brand:
            return KozmosColors.componentsPrimaryButtonsThemedButtonBackgroundIdle
        case .destructive:
            return KozmosColors.componentsPrimaryButtonsDangerButtonBackgroundIdle
        case .inverse:
            return KozmosColors.primitivesColorsBackground0
        }
    }

    private var foregroundColor: Color {
        switch tone {
        case .neutral:
            return KozmosColors.componentsPrimaryButtonsNeutralButtonForegroundContentIdle
        case .brand:
            return KozmosColors.componentsPrimaryButtonsThemedButtonForegroundContentIdle
        case .destructive:
            return KozmosColors.componentsPrimaryButtonsDangerButtonForegroundContentIdle
        case .inverse:
            return KozmosColors.primitivesColorsForeground100
        }
    }
}
